boost-maven/boost-maven-plugin/src/main/java/io/openliberty/boost/maven/plugin/AbstractMojo.java

```diff
@@ -88,31 +99,34 @@ public void execute() throws MojoExecutionException {
     }
     
     protected RuntimeI getRuntimeInstance() throws MojoExecutionException {
-        if(runtime == null) {
+        if (runtime == null) {
             BoostLogger logger = BoostLogger.getInstance();
-            if (dependencies.containsKey(AbstractBoosterConfig.RUNTIMES_GROUP_ID + ":tomee")) {
-                logger.info("Detected TomEE as target Boost runtime");
-                String tomeeInstallDir = projectBuildDir + "/apache-tomee/";
-                String tomeeConfigDir = tomeeInstallDir + "conf";
-                runtime = new TomeeRuntime(dependencies, getExecutionEnvironment(), tomeeInstallDir, tomeeConfigDir, getMavenDependencyPlugin());
-            } else if (dependencies.containsKey(AbstractBoosterConfig.RUNTIMES_GROUP_ID + ":openliberty")) {
-                logger.info("Detected Open Liberty as target Boost runtime");
-                String runtimeGroupId = "io.openliberty";
-                String runtimeArtifactId = "openliberty-runtime";
-                String runtimeVersion = "19.0.0.3";
-                runtime = new LibertyRuntime(dependencies, getExecutionEnvironment(), project, getLog(), repoSystem, repoSession, remoteRepos, getMavenDependencyPlugin(), runtimeGroupId, runtimeArtifactId, runtimeVersion);
-            } else if (dependencies.containsKey(AbstractBoosterConfig.RUNTIMES_GROUP_ID + ":wlp")) {
-                logger.info("Detected WebSphere Liberty as target Boost runtime");
-                String runtimeGroupId = "com.ibm.websphere.appserver.runtime";
-                String runtimeArtifactId = "wlp-javaee7";
-                String runtimeVersion = "19.0.0.3";
-                runtime = new LibertyRuntime(dependencies, getExecutionEnvironment(), project, getLog(), repoSystem, repoSession, remoteRepos, getMavenDependencyPlugin(), runtimeGroupId, runtimeArtifactId, runtimeVersion);
-            }
-            else {
-                throw new MojoExecutionException("No target Boost runtime was detected. Please add a runtime and restart the build.");
+            RuntimeParams params = new RuntimeParams(dependencies, getExecutionEnvironment(), project, getLog(), repoSystem, repoSession, remoteRepos, getMavenDependencyPlugin());
+            try {
```

Comment:
This RuntimeParams looks too complicated.

What I think we do is:  instead of the implementation using **ServiceLoader** to plug-in a Runtime impl, they plug-in a RuntimeFactory impl... so we have two interfaces, RuntimeFactoryI and RuntimeI.     The factory interface is the signature implied by:    `RuntimeI  createRuntime(dependencies, getExecutionEnvironment(), project, getLog())`   (I know that's not my lazy shorthand for listing the actual types).      This assumes that we only need the other stuff:  `repoSystem, repoSession, remoteRepos, ` to get the dependencies?   The maven dependency plugin is something an impl can get itself.

Would that work or am I missing something?

Reply:
Actually, let me amend that.    A factory interface is better than a constructor with **MyImpl(RuntimeParms)**.. but we could collapose this into one interface where **RuntimeI**   does `doPackage(dependencies, getExecutionEnvironment(), project, getLog())`.   I don't think there's any need to try to store "state" for start/stop, etc... so that'd be better than having two interfaces to implement (RuntimeI & RuntimeFactoryI).